### PR TITLE
Exercise4

### PR DESCRIPTION
**Pull request by SarahLM**
please review our exercise. 
this is the link to the review we gave: https://github.com/Systems-Development-and-Frameworks/no-js-pls/pull/6#pullrequestreview-547931835

### CODE REVIEW

File: backend/src/resolvers.js [hunk 1]
```diff
@@ -12,78 +14,96 @@ export default {
...
         });
       }
 
-      const newPost = {
-        title: args.title,
-        authorid: args.userId,
-        votes: 0,
-      };
+      // get User
+      const user = context.dataSources.db.getUserByEmail(email);
```
Comment (RespectableRuessel) [inline backend/src/resolvers.js:28]: Email is not validated. It is possible to use an empty string as email address. Possible fix is looking for the '@' symbol or doing a validation check.

File: backend/src/resolvers.js [hunk 2]
```diff
@@ -12,78 +14,96 @@ export default {
...
+        throw new UserInputError('400', {
+          invalidArgs:
+            '[Validation Errors] password is shorter then 8 characters',
+        });
       }
 
-      return context.dataSources.db.upvotePost(post.id, voter.id);
-    },
+      // try to find if the email already exist
+      const userExist = context.dataSources.db.getUserByEmail(email);
```
Comment (RespectableRuessel) [inline backend/src/resolvers.js:60]: Email is not validated. It is possible to use an empty string as email address. Possible fix is looking for the '@' symbol or doing a validation check.

File: backend/src/resolvers.js [hunk 3]
```diff
@@ -12,78 +14,96 @@ export default {
...
-    createPost: (parent, args, context) => {
-      // throw error if user does not exist
-      const user = context.dataSources.db.getUser(args.userId);
-      if (!user) {
-        throw new UserInputError('Invalid user', {
-          invalidArgs: 'Userid does not exist',
+    login: (parent, args, context) => {
+      const { email, password } = args;
+
+      if (password.length < 8) {
```
Comment (RespectableRuessel) [inline backend/src/resolvers.js:20]: Due to missing type checking it would be possible to pass an array (which also has a `length` method) to the args which will result in a true statement.

File: backend/src/utils/jwt.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    process.env.JWT_KEY || LOCAL_JWT_KEY,
+    { algorithm: 'HS256' }
+  );
+};
+
+export const verifyToken = (token) => {
+  const cleanedUpToken = token.startsWith('Bearer')
+    ? token.replace('Bearer ', '')
+    : token;
+  return jwt.verify(cleanedUpToken, process.env.JWT_KEY || LOCAL_JWT_KEY, {
```
Comment (RespectableRuessel) [inline backend/src/utils/jwt.js:20]: Maybe try to use different environment files with `dotenv -e FILE`.
Here you have an unnecessary check every time a token gets created or verified.

**Review by RespectableRuessel — APPROVED** [review]
Our review:
🟥 Missing rebase of homework/main
🟥 Add edge cases to the tests for the signup process of a user

:star: For an implemented signup feature.
:star: For a software-tested signup feature. **(see comment above)**
:star: For password validation **(maybe add a type check)**
:star: For unique email address validation.
:star: For not having security issues according to instructions.
:star: For a software-tested login feature.
:star: For assigning the authenticated user to a post in write and upvote mutations.
:star: For requesting a review and reviewing another team's PR according to the instructions.

**Review by medizinmensch — APPROVED** [review]
Hey!

Looks good!

You got :star: see below.

To be honest, I am a bit confused with your git graph. It looks like you did a rebase, but kept on pushing on the exercise4/exercise5 branches, can you confirm that?

I see you got a review, but I can't see which other group you reviewed.

So far 8/10 ⭐ 

EDIT: my gif is not working?!

**Comment by Herzogin** [comment]
Hej!

Wir haben einen Rebase gemacht und danach nur noch die Readme angepasst. Außerdem hatten wir schon einen Branch für Exercise 5 erstellt - da war aber noch nix drauf und ich hab ihn jetzt gelöscht. Jetzt sieht alles super ordentlich aus und du kannst uns unseren wohlverdienten Stern für das Rebase ruhig geben ;-)

Das Review hatten wir auch gemacht und ihn in unsere Readme geschrieben. Jetzt  steht er nochmal oben im Pull request - und ganz bequem - noch mal hier: [https://github.com/Systems-Development-and-Frameworks/no-js-pls/pull/6#pullrequestreview-547931835](url)   Den Stern für das Review hätten wir also bitte auch noch gerne :-)

Dankeschön!

> Hey!
> 
> Thanks to @RespectableRuessel ! I can only confirm: It looks good!
> 
> You got star see below plus the one for the security star
> 
> To be honest, I am a bit confused with your git graph. It looks like you did a rebase, but kept on pushing on the exercise4/exercise5 branches, can you confirm that?
> 
> I see you got a review, but I can't see which other group you reviewed.
> 
> So far 8/10 star
> 
> EDIT: my gif is not working?!



> Hey!
> 
> Thanks to @RespectableRuessel ! I can only confirm: It looks good!
> 
> You got star see below plus the one for the security star
> 
> To be honest, I am a bit confused with your git graph. It looks like you did a rebase, but kept on pushing on the exercise4/exercise5 branches, can you confirm that?
> 
> I see you got a review, but I can't see which other group you reviewed.
> 
> So far 8/10 star
> 
> EDIT: my gif is not working?!

**Comment by roschaefer** [comment]
> Hej!
> 
> Wir haben einen Rebase gemacht und danach nur noch die Readme angepasst. Außerdem hatten wir schon einen Branch für Exercise 5 erstellt - da war aber noch nix drauf und ich hab ihn jetzt gelöscht. Jetzt sieht alles super ordentlich aus und du kannst uns unseren wohlverdienten Stern für das Rebase ruhig geben ;-)
> 
> Das Review hatten wir auch gemacht und ihn in unsere Readme geschrieben. Jetzt steht er nochmal oben im Pull request - und ganz bequem - noch mal hier: [Systems-Development-and-Frameworks/no-js-pls#6 (review)](https://github.com/Systems-Development-and-Frameworks/no-js-pls/pull/6#pullrequestreview-547931835) Den Stern für das Review hätten wir also bitte auch noch gerne :-)
> 
> Dankeschön!
> 
> > Hey!
> > Thanks to @RespectableRuessel ! I can only confirm: It looks good!
> > You got star see below plus the one for the security star
> > To be honest, I am a bit confused with your git graph. It looks like you did a rebase, but kept on pushing on the exercise4/exercise5 branches, can you confirm that?
> > I see you got a review, but I can't see which other group you reviewed.
> > So far 8/10 star
> > EDIT: my gif is not working?!

Yes! I can confirm @Herzogin so you get all 10/10 :star:.
* I fixed the link to the Review
* I checked that the rebase looks right
